### PR TITLE
Add deque to kstd

### PR DESCRIPTION
**Pull request by Cach30verfl0w**

### CODE REVIEW

File: include/kstd/Deque.hpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    private:
+        class Iterator final {
+            node_type* _current_node;
+
+        public:
+            explicit Iterator(node_type* current_node) noexcept
+                : _current_node(current_node) {
+            }
+
+            auto operator==(const Iterator& other) const noexcept -> bool {
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:44]: comparison operator should be `[[nodiscard]]` in my opinion

File: include/kstd/Deque.hpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            node_type* node = _first_node;
+            for (usize i = 0; i < index - 1; i++) {
+                node = node->next;
+            }
+
+            // Create new node
+            node_type* new_node = _allocator.allocate(1);
+            new_node->last = node;
+            new_node->next = node->next;
+            new_node->data = value;
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:109]: please use `new_node->data = kstd::move(value)` here

File: include/kstd/Deque.hpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            if (node->next != nullptr) {
+                node->next->last = new_node;
+            } else {
+                _last_node = new_node;
+            }
+            node->next = new_node;
+            _size++;
+        }
+
+        auto push_front(const T data) noexcept -> void {
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:121]: please use consistent naming here. Name the parameter `value` like in `insert`

File: include/kstd/Deque.hpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            node_type* node = _allocator.allocate(1);
+            new(node) node_type();
+            node->data = data;
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:124]: `kstd::move(data)`

File: include/kstd/Deque.hpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            }
+            _size++;
+        }
+
+        auto pop_front() noexcept -> T {
+            if(_first_node == nullptr) {
+                panic("Unable to pop front because no elements are inserted");
+            }
+
+            auto value = _first_node->data;
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:141]: we should also move out here

File: include/kstd/Deque.hpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        }
+
+        auto peek_front() noexcept -> T& {
+            if(_first_node == nullptr) {
+                panic("Unable to pop front because no elements are inserted");
+            }
+            return _first_node->data;
+        }
+
+        auto push_back(const T data) noexcept -> void {
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:160]: `data` -> `value`

File: include/kstd/Deque.hpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            node_type* node = _allocator.allocate(1);
+            new(node) node_type();
+            node->data = data;
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:163]: `kstd::move`

File: include/kstd/Deque.hpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            }
+            _size++;
+        }
+
+        auto pop_back() noexcept -> T {
+            if(_last_node == nullptr) {
+                panic("Unable to pop back because no elements are inserted");
+            }
+
+            auto value = _last_node->data;
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:180]: also move out here

**Review by BeastLe9enD — CHANGES_REQUESTED** [review]
Looks fine! Just a few `kstd::move` / naming things 🙂

**Comment by KitsuneAlex** [comment]
Very nice :)

File: include/kstd/Deque.hpp [hunk 9]
```diff
@@ -177,9 +177,9 @@ namespace kstd {
                 panic("Unable to pop back because no elements are inserted");
             }
 
-            auto value = _last_node->data;
+            auto value = std::move(_last_node->data);
             _last_node = _last_node->last;
-            return value;
+            return std::move(value);
```
Comment (BeastLe9enD) [inline include/kstd/Deque.hpp:182]: move on return semantics

**Review by KitsuneAlex — APPROVED** [review]
Looks good to me

**Pull request merged**